indexer/sql/src/non_blocking_indexer.rs

```diff
@@ -0,0 +1,343 @@
+            indexing: false,
+        })
+    }
+}
+
+/// Because I'm using `.spawn` in this implementation, I ran into lifetime issues where I need the
+/// data to live as long as the spawned task.
+///
+/// I also have potential issues where the task spawned in `pre_indexing` to create a DB
+/// transaction (in the sync implentation of this trait) could be theorically executed after the
+/// task spawned in `index_block` and `index_transaction` meaning I'd have to check in these
+/// functions if the transaction exists or not.
+///
+/// Decided to do different and prepare the data in memory to inject all data in a single Tokio
+/// spawned task
+#[derive(Debug, Clone)]
+pub struct NonBlockingIndexer {
+    pub context: Context,
+    pub runtime: Option<Arc<Runtime>>,
+    pub handle: tokio::runtime::Handle,
+    blocks: Arc<Mutex<HashMap<u64, BlockToIndex>>>,
+    // NOTE: this could be Arc<AtomicBool> because if this Indexer is cloned all instances should
+    // be stopping when the program is stopped, but then it adds a lot of boilerplate. So far, code
+    // as I understand it doesn't clone `App` in a way it'd raise concern.
+    indexing: bool,
+}
+
+/// Saves the block and its transactions in memory
+#[derive(Debug, Clone)]
+struct BlockToIndex {
+    pub block_info: BlockInfo,
+    pub txs: Vec<(Tx, TxOutcome)>,
+}
+
+impl BlockToIndex {
+    /// Takes care of inserting the data in the database
+    pub async fn save(self, db: &DatabaseTransaction) -> error::Result<()> {
+        #[cfg(feature = "tracing")]
+        tracing::info!(block_height = self.block_info.height, "Indexing block");
+
+        let mut models = Models::build(&self.block_info)?;
+        for tx in self.txs.iter() {
+            models.push(&tx.0, &tx.1)?;
+        }
+
+        // TODO: if the process was to crash in the middle and restarted, we could try to
+        // reinsert existing data. We should use `on_conflict()` to avoid this, return the
+        // existing block and change `block_id` when/if we added foreign keys
+        models.block.insert(db).await?;
+        entity::transactions::Entity::insert_many(models.transactions)
+            .exec(db)
+            .await?;
+        entity::messages::Entity::insert_many(models.messages)
+            .exec(db)
+            .await?;
+        entity::events::Entity::insert_many(models.events)
+            .exec(db)
+            .await?;
+        Ok(())
+    }
+}
+
+impl NonBlockingIndexer {
+    fn find_or_create<F, R>(&self, block: &BlockInfo, action: F) -> error::Result<R>
+    where
+        F: FnOnce(&mut BlockToIndex) -> error::Result<R>,
+    {
+        let mut blocks = self.blocks.lock().expect("Can't lock blocks");
+        let block_to_index = blocks.entry(block.height).or_insert(BlockToIndex {
+            block_info: *block,
+            txs: vec![],
+        });
+        action(block_to_index)
+    }
+
+    fn find_or_fail(&self, block_height: &u64) -> error::Result<BlockToIndex> {
```

Comment:
`u64` implement `Copy`, don't borrow it
```suggestion
    fn find_or_fail(&self, block_height: u64) -> error::Result<BlockToIndex> {
```

Reply:
How do you return a `&BlockToIndex` which is a ref to a `Arc<Mutex<HashMap>>`?